### PR TITLE
[Merged by Bors] - feat: presented monoids

### PR DESCRIPTION
**Pull request by hannahfechtner**
Define presented monoids and their universal property. This requires some additions to free monoids, and GroupTheory/Congruence

---

- [ ] depends on: #17180

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:223]: This is a general result that If `e : M →* N` is surjective then `(c.comap e).Quotient ≃* c.Quotient` with `c : Con N`.

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:36]: Can you add 
```
lemma eqvGen_stepsto {β : Type*} [Monoid β] (r : β → β → Prop) :
    EqvGen (PresentedMonoid.StepsTo r) = ConGen.Rel r := by
``` 
`PresentedMonoid.StepsTo` should have a different name (maybe `MulGen`?) and be moved to `GroupTheory/Congruence`

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:40]: I still think `abbrev PresentedMonoid := (conGen rel).Quotient` is a better definition. It gives you the monoid structure and the surjection `FreeMonoid α -> PresentedMonoid rel` and the universal property for free.

**Review by erdOne — COMMENTED** [review]
Using existing material to simplify your proof is only half the story. Sometimes it is the stuff you are adding that is improving the existing material (for example `(c.comap e).Quotient ≃* c.Quotient`). Also people might want to use results from both sides and it is useful to have a common backbone that is being built upon.

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: implemented in the latest PR!

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: now it's implemented with Con

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: done!

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (rosborn) [inline Mathlib/Algebra/FreeMonoid/Basic.lean:429]: I'm assuming this should be deleted?

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (rosborn) [inline Mathlib/Algebra/FreeMonoid/Basic.lean:480]:
```suggestion
  rcases fs with ⟨a, rfl⟩
  rcases ih with ⟨b, rfl⟩
  use FreeMonoid.of a * b
```

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (rosborn) [inline Mathlib/Algebra/FreeMonoid/Basic.lean:504]: I'm curious what the use case of this would be.

Comment (hannahfechtner) [reply Mathlib/Algebra/FreeMonoid/Basic.lean:504]: It might be overkill to use foldr; perhaps there's a more specific way. In a later file, I'll need to embed into a group. So I want to multiply together all the symbols in the free monoid element, using the group's multiplication

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/FreeMonoid/Basic.lean]: oh oops! that and the one below it, too

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (rosborn) [reply Mathlib/Algebra/FreeMonoid/Basic.lean:499]: Sorry for the late response, but that is just [FreeMonoid.lift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/FreeMonoid/Basic.html#FreeMonoid.lift). The reason, I am suspicious of `foldr` is that I don't think we should just duplicate the `List` API in `FreeMonoid`.

Comment (hannahfechtner) [reply Mathlib/Algebra/FreeMonoid/Basic.lean:499]: oh, I see! that makes sense, I'll remove it

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (erdOne) [inline Mathlib/GroupTheory/Congruence/Basic.lean:1162]:
```suggestion
noncomputable def comapQuotientEquivOfSurj (c : Con M) (f : N →* M) (hf : Function.Surjective f) :
```

File: Mathlib/Control/Fold.lean
Comment (erdOne) [reply Mathlib/Control/Fold.lean:341]: What happened here?

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (erdOne) [reply Mathlib/Algebra/FreeMonoid/Basic.lean:499]: What's the status of this?

**Review by erdOne — COMMENTED** [review]
LGTM modulo minor comments

**Comment by erdOne** [comment]
Also please merge master in case of merge conflicts. Thanks!

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/FreeMonoid/Basic.lean]: it's deleted as of the last commit

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (YaelDillies) [inline Mathlib/Algebra/FreeMonoid/Basic.lean:1]: Could this be a separate PR?

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:1]: And this too?

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:565]:
```suggestion
theorem comap_conGen_of_bijective {M N : Type*} [Mul M] [Mul N] (f : M → N)
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (YaelDillies) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:1]: This could become the main new thing of this PR

**Comment by YaelDillies** [comment]
I think people who are not interested in presented monoids might be interested in the parts I'm asking you to split off.

File: Mathlib/Algebra/FreeMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/FreeMonoid/Basic.lean:1]: it is now PR#17180

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:120]:
```suggestion
def toMonoid : PresentedMonoid rels →* M :=
```
I would call this PresentedMonoid.lift instead.

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:114]:
```suggestion
variable (h : ∀ a b : FreeMonoid α, rels a b → FreeMonoid.lift f a = FreeMonoid.lift f b)
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:137]: This is easier proved via [MonoidHom.eq_of_eqOn_denseM](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Submonoid/Basic.html#MonoidHom.eq_of_eqOn_denseM)

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:129]:
```suggestion
theorem toMonoid.of {x : α} : toMonoid f h (of rels x) =
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:56]:
```suggestion
  mk rels (.of x)
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:48]: If you move the monoid instance on `PresentedMonoid` before this, this can be a monoid hom instead.

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: I think `lift` makes more sense here (or even `desc`).

Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: done!

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: fixed!

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: done :)

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean:1]: this is now PR#17195

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:19]: Is "steps-to" closure a standard terminology? I've never heard of it.

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:21]:
```suggestion
* `PresentedMonoid.of`: The canonical map from `α` to a presented monoid with generators `α`.
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:22]:
```suggestion
* `PresentedMonoid.lift f`: the canonical monoid homomorphism `PresentedMonoid rels → M`, given a function
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:66]:
```suggestion
variable {α₁ α₂ α₃ : Type*} {rels₁ : FreeMonoid α₁ → FreeMonoid α₁ → Prop}
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:95]:
```suggestion
additive monoid. That is, the additive submonoid closure of the set of generators equals `⊤`"]
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean]: Oy, it is not! I went back to some textbooks, and it looks like the proper term is "one-step reduction"

I've updated the definitions (thanks to your comment, I realized I had other errors too!), though it feels clunky. I'm not sure how to best describe `rels`. It's not a set - can I just call it a relation? Or is it a function? Any advice would be much appreciated!

(I may also have better ideas in the morning)

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:127]:
```suggestion
theorem lift_of {x : α} : lift f h (of rels x) = f x := rfl
```

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (erdOne) [inline Mathlib/Algebra/PresentedMonoid/Basic.lean:52]: I think `surjective_mk` would be useful too.

**Review by erdOne — APPROVED** [review]
LGTM modulo minor comments

File: Mathlib/Algebra/PresentedMonoid/Basic.lean
Comment (hannahfechtner) [reply Mathlib/Algebra/PresentedMonoid/Basic.lean:52]: added! it's a bit below, because the proof uses PresentedMonoid.inductionOn

**Comment by erdOne** [comment]
LGTM Thanks!
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Review by Vierkantor — APPROVED** [review]
Thanks!

bors r+

**Comment by Vierkantor** [comment]
Oops, I could swear I reloaded this page before `bors`ing. Anyway, I'll give this a

bors d+

in case the build ends up failing.

**Pull request closed without merging**
